### DISCUSSION TITLE
Closing connections. Migration 1.4 to 2.*

### DISCUSSION THREAD

**Discussion by FranckRodriguez** [sqlalchemy/sqlalchemy#11835] [Usage Questions]
Hello, 

I'm currently trying to upgrade SQLAlchemy from 1.4 to 2.0 on a pretty huge project using python 3.10.9 and SQLite database

I use this class for all my accesses to database (queries, insertions, ...):
```py
class Repo():
    def __init__(self, path_to_DB):
        self.engine = create_engine("sqlite:///" + path_to_DB, echo=False, future=True)
        with self.engine.connect() as conn:
            conn.execute(text("PRAGMA foreign_keys=ON"))

    def commit(self):
        with Session(self.engine) as session:
            session.commit()

    def insert(self, *objs):
        with Session(self.engine) as session:
            session.add_all(objs)
            session.commit()
```
and much, much more functions in child classes...

When using pytest to run tests using pytest fixtures like :
```py
@pytest.fixture
def create_new_DB():
    try:
        os.remove(path_to_DB)
    except Exception as e:
        print(e)
    create_or_load_DB(path_to_DB, fill=True, fill_user=False)  # create DB with all tables without users info
```
The second call to the fixture fails to remove database because "used by another process" (the previous test, or the test session i guess).

I have this issue only with SQLAlchemy 2.*.

I think its obviously connections closing changes, but idk how I can fix it (I've already spent weeks on this topic)

If someone has an idea to help me....

PS: I'm not a crack, and some conecpts are abstract to me, so feel free to talk to me as a no brained child !

**Comment by zzzeek** [comment]
switch to NullPool explicitly for SQLite so that there isn't a pooled connection present

https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#the-sqlite-dialect-uses-queuepool-for-file-based-databases

**Reply by FranckRodriguez** [reply]
It says that it "did in fact have a measurable negative performance impact". Do I understand well? I guess i should check what is a pooled connection to avoid bad consequences on the project....

**Reply by zzzeek** [reply]
your first goal is to get your code working fully under the new version.    The NullPool / QueuePool thing is a very minimal performance impact which you will not notice

**Reply by CaselIT** [reply]
If you want to try keeping the queue pool, you should call `engine.dispose()` before `os.remove(path_to_DB)` to close all existing connections

**Reply by FranckRodriguez** [reply]
OK, thx. I will try those ideas.

In an other project, my predecessor didn't used any with statement, and created a new session for almost all queries (+200 different ones). I'm trying to transform it like my example in my first message, with only with statements.

I have the same problem for my pytest sessions and tried all engine.dispose, session.close, ...  found in the sqlalchemy doc.
I also tried the NullPool on it.

Is there anything else i can try to kill every link with the database between 2 tests to be able to continue my tests before finishing queries transformations?
